### PR TITLE
Change the signature of signup/signin methods

### PR DESCRIPTION
**Pull request by SaurabhMSFT**
The signIn and signInUsingPassword methods have been refactored into one signIn method with optional password. 
The signUp and signUpUsingPassword methods have been refactored into one signUp method with optional password. 

PR for the common repo is https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2284

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/nativeauth/INativeAuthPublicClientApplication.kt [hunk 1]
```diff
@@ -71,89 +69,47 @@ interface INativeAuthPublicClientApplication : IPublicClientApplication {
...
      * @throws MsalClientException if an account is already signed in.
      */
-    suspend fun signUp(username: String, attributes: UserAttributes? = null): SignUpResult
+    suspend fun signUp(username: String, password: CharArray? = null, attributes: UserAttributes? = null): SignUpResult
 
     /**
      * Sign up the account starting from a username; callback variant.
      *
      * @param username username of the account to sign up.
+     * @param password password of the account to sign up.
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/nativeauth/INativeAuthPublicClientApplication.kt]: let's make it explicit that this is optional, like we do for attributes.

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt [hunk 2]
```diff
@@ -414,22 +280,24 @@ class NativeAuthPublicClientApplication(
...
         scopes: List<String>?
-    ): SignInUsingPasswordResult {
+    ): SignInResult {
         LogSession.logMethodCall(TAG, "${TAG}.signInUsingPassword")
         return withContext(Dispatchers.IO) {
             LogSession.logMethodCall(TAG, "${TAG}.signInUsingPassword.withContext")
 
             verifyNoUserIsSignedIn()
 
+            val hasPassword = password != null && !password.isEmpty()
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt:299]: `password.isNullOrBlank()`

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt [hunk 3]
```diff
@@ -414,22 +280,24 @@ class NativeAuthPublicClientApplication(
...
+    override suspend fun signIn(
         username: String,
-        password: CharArray,
+        password: CharArray?,
         scopes: List<String>?
-    ): SignInUsingPasswordResult {
+    ): SignInResult {
         LogSession.logMethodCall(TAG, "${TAG}.signInUsingPassword")
         return withContext(Dispatchers.IO) {
             LogSession.logMethodCall(TAG, "${TAG}.signInUsingPassword.withContext")
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: Update log (and perhaps also in other places)

File: msal/src/main/java/com/microsoft/identity/nativeauth/INativeAuthPublicClientApplication.kt (hunk 1 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/INativeAuthPublicClientApplication.kt]: done

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt (hunk 3 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: done

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java [hunk 4]
```diff
@@ -413,16 +374,16 @@ public static SignUpStartCommandParameters createSignUpStartCommandParameters(
      * @return Command parameter object
      * @throws ClientException
      */
-    public static SignUpStartUsingPasswordCommandParameters createSignUpStartUsingPasswordCommandParameters(
+    public static SignUpStartCommandParameters createSignUpStartCommandParameters(
             @NonNull final NativeAuthPublicClientApplicationConfiguration configuration,
             @NonNull final OAuth2TokenCache tokenCache,
             @NonNull final String username,
-            @NonNull final char[] password,
+            @Nullable final char[] password,
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:381]: update comment (line 373) to reflect the fact that password is optional

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java [hunk 5]
```diff
@@ -634,11 +554,11 @@ public static SignInStartCommandParameters createSignInStartCommandParameters(
      * @return Command parameter object
      * @throws ClientException
      */
-    public static SignInStartUsingPasswordCommandParameters createSignInStartUsingPasswordCommandParameters(
+    public static SignInStartCommandParameters createSignInStartCommandParameters(
             @NonNull final NativeAuthPublicClientApplicationConfiguration configuration,
             @NonNull final OAuth2TokenCache tokenCache,
             @NonNull final String username,
-            @NonNull final char[] password,
+            @Nullable final char[] password,
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:561]: Same - update comment

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt [hunk 6]
```diff
@@ -588,18 +492,18 @@ class NativeAuthPublicClientApplication(
...
 
             val parameters =
-                CommandParametersAdapter.createSignUpStartUsingPasswordCommandParameters(
+                CommandParametersAdapter.createSignUpStartCommandParameters(
                     nativeAuthConfig,
                     nativeAuthConfig.oAuth2TokenCache,
                     username,
                     password,
-                    attributes?.toMap()
+                    attributes?.userAttributes
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: why this change?

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java [hunk 7]
```diff
@@ -2587,17 +2582,17 @@ public void testSignUpInvalidPasswordReturnsError() throws ExecutionException, I
                 MockApiResponseType.PASSWORD_TOO_WEAK
         );
 
-        SignUpUsingPasswordTestCallback signUpUsingPasswordTestCallback = new SignUpUsingPasswordTestCallback();
-        application.signUpUsingPassword(username, password, UserAttributes.Builder.customAttribute("password", password.toString()).build(), signUpUsingPasswordTestCallback);
+        SignUpTestCallback signUpTestCallback = new SignUpTestCallback();
+        application.signUp(username, password, UserAttributes.Builder.customAttribute("password", password.toString()).build(), signUpTestCallback);
```
Comment (SammyO) [inline msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java]: I never saw this before, but I find this strange. Why are we passing password as a custom attribute? It shouldn't matter technically, but we should remove it to avoid confusion.

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java [hunk 8]
```diff
@@ -2606,13 +2601,13 @@ public void testSignUpUsingPasswordInvalidEmailReturnsError() throws ExecutionEx
                 MockApiResponseType.INVALID_USERNAME
         );
 
-        SignUpUsingPasswordTestCallback signUpUsingPasswordTestCallback = new SignUpUsingPasswordTestCallback();
-        application.signUpUsingPassword(invalidUsername, password, UserAttributes.Builder.customAttribute("password", password.toString()).build(), signUpUsingPasswordTestCallback);
+        SignUpTestCallback signUpTestCallback = new SignUpTestCallback();
+        application.signUp(invalidUsername, password, UserAttributes.Builder.customAttribute("password", password.toString()).build(), signUpTestCallback);
```
Comment (SammyO) [inline msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java]: Same here.

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java [hunk 9]
```diff
@@ -2626,7 +2621,7 @@ public void testSignUpInvalidEmailReturnsError() throws ExecutionException, Inte
         );
 
         SignUpTestCallback signUpTestCallback = new SignUpTestCallback();
-        application.signUp(invalidUsername, UserAttributes.Builder.customAttribute("password", password.toString()).build(), signUpTestCallback);
+        application.signUp(invalidUsername, null, UserAttributes.Builder.customAttribute("password", password.toString()).build(), signUpTestCallback);
```
Comment (SammyO) [inline msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java]: And here

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java (hunk 4 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:381]: The password is nullable but is still required as Java does not support default values. Developers have to specify all parameters.

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java (hunk 5 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:561]: The password is nullable but is still required as Java does not support default values. Developers have to specify all parameters.

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java (hunk 4 above)
Comment (SammyO) [reply msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:381]: This class is only used internally, but Kotlin classes (our state machine). In those cases, this field is optional. We've done the same for e.g. `scopes` parameter in other methods.

**Review by GBakolas — APPROVED** [review]

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt [hunk 10]
```diff
@@ -414,24 +280,26 @@ class NativeAuthPublicClientApplication(
...
-        LogSession.logMethodCall(TAG, "${TAG}.signInUsingPassword")
+    ): SignInResult {
+        LogSession.logMethodCall(TAG, "${TAG}.signIn")
         return withContext(Dispatchers.IO) {
-            LogSession.logMethodCall(TAG, "${TAG}.signInUsingPassword.withContext")
+            LogSession.logMethodCall(TAG, "${TAG}.signIn.withContext")
 
             verifyNoUserIsSignedIn()
 
+            val hasPassword = password != null && !password.isEmpty()
```
Comment (Yuki-YuXin) [inline msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: Maybe password.isNotEmpty()

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt [hunk 11]
```diff
@@ -568,15 +470,17 @@ class NativeAuthPublicClientApplication(
...
         username: String,
-        password: CharArray,
+        password: CharArray?,
         attributes: UserAttributes?
-    ): SignUpUsingPasswordResult {
-        LogSession.logMethodCall(TAG, "${TAG}.signUpUsingPassword(username: String, password: String, attributes: UserAttributes?)")
+    ): SignUpResult {
+        LogSession.logMethodCall(TAG, "${TAG}.signUp")
+
+        var hasPassword = password != null && !password.isEmpty()
```
Comment (Yuki-YuXin) [inline msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: The same here.

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt (hunk 6 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: This is what it is on line 781 in the previous version of the file.

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt (hunk 11 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: Done

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt (hunk 10 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: Done

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java (hunk 9 above)
Comment (SaurabhMSFT) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java]: Done

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java (hunk 8 above)
Comment (SaurabhMSFT) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java]: Done

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java (hunk 7 above)
Comment (SaurabhMSFT) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java]: Done

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java (hunk 4 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:381]: Java does not have optional parameters. Optional parameters are simulated in Java by overloading functions. As this parameter is in the middle, it wouldn't be good style to overload function with reduced set of parameters.

**Review by Yuki-YuXin — APPROVED** [review]

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt (hunk 6 above)
Comment (SammyO) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: It seems we had an inconsistency between `signUp` and `signUpUsingPassword` when it comes to this field.
Could you validate that `attributes?.userAttributes` is the right way of doing this, and not `attributes?.toMap()`, by doing some E2E testing.

Comment (SammyO) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: `SignUpStates.submitAttribute` still uses `toMap()`, so this inconsistency needs to be fixed. Let's make sure we're not introducing a last minute bug in this flow.

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt [hunk 12]
```diff
@@ -414,24 +280,26 @@ class NativeAuthPublicClientApplication(
      * Sign in the account using username and password; Kotlin coroutines variant.
      *
      * @param username username of the account to sign in.
-     * @param password password of the account to sign in.
+     * @param password (Optional) password of the account to sign in.
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt:283]: This needs to be removed, as the optionality doesn't apply to Java call sites. Applies in different places in the code.

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java (hunk 4 above)
Comment (SammyO) [reply msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:381]: My point is that this method isn't called from any Java class, and so we don't have to consider the non-optionality of this parameter. You can leave it like this however, that's also fine.

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt (hunk 6 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: `attributes?.userAttributes` and `attributes?.toMap()` return the same value. I have changed the usage here to `attributes?.toMap()`

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt (hunk 12 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt:283]: The place where the optionality doesnot apply is in an internal package and not used from the Sample app. Optional in the doc should stay here as it provides meaningful information to the developer.

**Review by SammyO — APPROVED** [review]
